### ISSUE TITLE
Position refresh repo ideas

### ISSUE THREAD

**Issue by licaon-kter** [LonamiWebs/Stringlate#84]
I find it pretty easy to hit by mistake, eg. when aiming for the search icon.
Yes, you get a dialogue but I'm afraid I might press "do not keep local changes" and loose all my work.

Maybe this can be moved in the menu?

**Comment by XyLoNaMiyX** [comment]
> but I'm afraid I might press "do not keep local changes" and loose all my work.

I was thinking about this recently and how I should clarify what this does. The only thing "Do not keep local changes does" is, _if**f**_ the repository _has **newer** strings_ for the ones you have translated, then _only those available_ will be updated. If the repository has no modification for your locale, _no changes will be lost_. But of course this is a lengthy and confusing explanation which wouldn't fit in a dialog button so I couldn't write it there.

> Maybe this can be moved in the menu?

It could, although it's not in the menu because, usually, you should check in case there are new strings on that repository. But yes, it doesn't happen that often and moving it to the menu would only make it be 2 taps away instead 1.

**Comment by XyLoNaMiyX** [comment]
> I was thinking about this recently and how I should clarify what this does. […]

I would still like some feedback @licaon-kter, if you could provide any about it, about this misunderstanding.

**Comment by licaon-kter** [comment]
Actually the dialogue is useless, might has well have it removed, there is no need for a user interaction:
* new string untranslated -> pull them
* old strings deleted -> delete local too
* old strings updated -> show another (read-only) field with `previous translation`, once a new translation is in place never show the old one

Simples. :)

**Comment by XyLoNaMiyX** [comment]
> old strings deleted -> delete local too

Not so simple. After all, they've put work into the translations. Even if there is currently no way to bring those strings back, *they still have translated them*. [I don't find it to be OK](https://github.com/LonamiWebs/Stringlate/blob/6bd13346369ee6d517fe719c4be50c71021e2bbb/src/app/src/main/java/io/github/lonamiwebs/stringlate/activities/translate/TranslateActivity.java#L838-L841) to just go and delete their work without any feedback.

> old strings updated -> show another

Not so simple, either. So, the work should be shown as "old translation" but not removed? How should it be handled later? Can it be removed manually? If so, when and how? What do you exactly mean by "*once a new translation is in place*"?

**Comment by licaon-kter** [comment]
> delete their work without any feedback

Ok, not sure what's the point, the user keeps the strings the project will never use anyway? Makes a collection? Prints them? Ok, I agree it's not nice, but you are at the will of the project maintainer (you like this project and developer enough to begin with, so you respect their decisions, right?), better then raise an issue in their project or something. Also, let's be honest, who is going to know. ;)

Oh, give some feedback: `7 strings deleted, 4 new, 3 updated. OK`

> "once a new translation is in place"?

This is the tricky part, might need another table for the project.
Basically, updated string gets no translation in the saved file.
When the user reaches a string that was translated in the past, then you show the old one, if the user adds a translation for that string (and saves it) then you clean old translation string from this additional table (or column or whatever structure you have there).

**Comment by XyLoNaMiyX** [comment]
> Oh, give some feedback: `7 strings deleted, 4 new, 3 updated. OK`

"*7 unused strings deleted, 4 new added, 3 updated*". That's a nice toast.

> might need another **table** for the project.

Except I don't have *tables* since I work on plain `.xml` files directly. So yeah, it's tricky.

> then you show the old one

And here we have the same problem again. Where should this be shown?

> if the user adds a translation for that string

"Adds"? There is already the old one, and the new pulled one. Which one should the user be modifying? "Keep changes"? "Discard old string"? But again keep changes and discard old string is what we have already. Show a `diff` letting the users decide which strings they should keep and which they want to remove (that won't be easy)?

**Comment by licaon-kter** [comment]
You create a new read-only field, so you'll have:
* Locale
* String id
* Original string
* *Old string* <- read only, but you can copy it (same issue as #2)
* Translated string <- ths one is either edited or empty (when exported I mean)

You only use XML? So, create a new tag, make old strings as `original_string_MYTAGG7GGG` (or something rather unique to avoid collisions) and you show that, delete it if new string gets translated, and except it from export.

Weblate iirc has diffs, they're nice but are they easy to implement? Diff colours, strikeout text, etc.

**Comment by XyLoNaMiyX** [comment]
> Old string <- read only, but you can copy it

Okay, so another field, say:
- original string (`ro`)
- translated old string (`ro`)
- translated updated string (`rw`).

But wouldn't that perhaps be a bit confusing for the users? "What does this «old string» thing mean? Why does it appear only sometimes?".

> Translated string <- ths one is either edited **or empty (when exported I mean)**

No, empty strings are not exported at all so it cannot be empty. It should be the updated string.

> So, create a new tag

I was thinking something like `old="true"` instead and keep both with the same ID. They don't have to be unique really (i.e. always a different ID), because that file is only for internal use.

> delete it if new string gets translated

The problem we're talking about is because we pulled a new, updated translation, different from the current one you had. If your translation got merged and you pulled that, there would be no problem because they would be the same. I think you meant "delete if new string gets **changed**"? But what if it's not changed, because it's *right* and the old one is *wrong*?

> Weblate iirc has diffs, they're nice but are they easy to implement? Diff colours, strikeout text, etc.

I don't think it would be *that* hard since we would only need to compare if two strings are equal or not to show "Different strings: Which one should be kept". Not sure what the `diff` would be useful for. Bear in mind also, despite sounding obvious, that phone screens are not as large as computer screens.

**Comment by licaon-kter** [comment]
> But wouldn't that perhaps be a bit confusing for the users?

Add more text `String updated, old translation was...` ?

> No, empty strings are not exported 

I know

> keep both with the same ID

I don't know how it can be made easier, if this works then great. My first thought was exactly this, but I guessed (wrong) that string IDs need to be unique..

> I think you meant "delete if new string gets changed"?

That's one case.

> But what if it's not changed, because it's right and the old one is wrong?

Not sure I follow.

There are 2 states, when having a string updated:
* string is translated again -> update translation
* string is NOT translated again -> NO translation, basically NO EXPORT for this string EVEN IF it was translated before.

I mean, the user NEEDS to copy old-to-new (using a button or like #2) EVEN if the translation remains unchanged, the user NEEDS to act upon this, to either write a new translation or CONFIRM the old was is still ok.

(if that makes sense)

**Comment by licaon-kter** [comment]
> "String updated, old translation was XXXX, new one is YYYY. Which one should be kept?"?.

Umm, no message... use the fields... if this string was updated then the NEW MIDDLE field appears, translation field is empty, user needs to do something, copy-old or write-new in that said translation field, else string is not translated, not exportable.

Maybe the first time it appears it will a bit confusing... but then s/he reads the header of that field ( `String was updated, old translation was:` ) and understands, and from now on, when s/he sees that field its purpose is known.

**Comment by Lonami** [comment]
[`ecac8a3`](https://github.com/LonamiWebs/Stringlate/commit/ecac8a3d2fd704af6e208fa859976b7826744010) sets up the ground by introducing the new text when updating strings, and [`d282fb8`](https://github.com/LonamiWebs/Stringlate/commit/d282fb8c08c60ae469b7200494c1c081e0884765) implements its behaviour. It should be clearer now. Unused strings are deleted always without asking the user, because they're not needed for the translation anymore. I don't think anyone would miss them.
